nix/docker.nix

```
@@ -0,0 +1,123 @@
+############################################################################
+# Docker image builder
+#
+# To test it out, use:
+#
+#   docker load -i $(nix-build -A dockerImage.base --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.latency-tests --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.mainnet --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.mainnet-ci --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.shelley_staging --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.shelley_staging_short --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.staging --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.testnet --no-out-link)
+#
+# For all output except base, run:
+#
+#   docker run inputoutput/cardano-node:<TAG>
+#
+# ie:
+#
+#   docker run inputoutput/cardano-node:6a5b233e5861aa0012287a5557b3a6a08afbf3ca-testnet
+#
+# For base, you will need to provide all variables. ie:
+#
+#   docker run -ti \
+#   --volume $PATH_TO/state-docker:/data \
+#   --volume $PATH_TO/configuration-mainnet.yaml:/config/config.yaml \
+#   --volume $PATH_TO/mainnet-topology.json:/config/topology.json \
+#   --volume $PATH_TO/mainnet-genesis.json:/config/genesis.json \
+#   inputoutput/cardano-node:<GIT_HASH>-custom
+#   --genesis-hash <GENESIS_HASH>
+#
+############################################################################
+
+{ system ? builtins.currentSystem
+, crossSystem ? null
+, config ? {}
+, profiling ? false
+, pkgs ? import ../nix { inherit system crossSystem config profiling; }
+, dockerTools
+
+# The main contents of the image.
+, cardano-node
+, scripts
+
+# Get the current commit
+, gitrev ? pkgs.iohkNix.commitIdFromGitRepoOrZero ../.git
+
+# Other things to include in the image.
+, bashInteractive
+, cacert
+, coreutils
+, curl
+, glibcLocales
+, iana-etc
+, iproute
+, iputils
+, socat
+, utillinux
+
+, repoName ? "inputoutput/cardano-node"
+}:
+
+with pkgs;
+let
+
+  # Layer of tools which aren't going to change much between versions.
+  baseImage = dockerTools.buildImage {
+    name = "${repoName}-env";
+    contents = [
+      bashInteractive   # Provide the BASH shell
+      cacert            # X.509 certificates of public CA's
+      coreutils         # Basic utilities expected in GNU OS's
+      curl              # CLI tool for transferring files via URLs
+      glibcLocales      # Locale information for the GNU C Library
+      iana-etc          # IANA protocol and port number assignments
+      iproute           # Utilities for controlling TCP/IP networking
+      iputils           # Useful utilities for Linux networking
+      socat             # Utility for bidirectional data transfer
+      utillinux         # System utilities for Linux
+    ];
+    # set up /tmp (override with TMPDIR variable)
+    extraCommands = ''
+      mkdir -m 0777 tmp
+    '';
+  };
+  customImage = dockerTools.buildImage {
+    name = "${repoName}";
+    tag = "${gitrev}-base";
```

```suggestion
    tag = "${gitrev}";
```
The base image needs no network specifier

---

that's why I thought custom was a better name. My concern is someone is going to run it with just the gitrev and then complain on social media our code sucks because they didn't read the instructions and that this image is strictly for debugging with local custom files. debug might be a good tag as well.

---

It's **not** strictly for debugging! It's simply the base image that wraps the `cardano-node` service, which has similar requirements for the user, so should not come as a suprise. Please consider separating the psychological attachment with nix being the only source of config